File: relay-ios/FaceChat/Sources/HudPresenter.swift
import Foundation
import MWDATCore
import MWDATDisplay

@MainActor
final class HudPresenter {
  enum State { case ambient, peeking, engaged, followup, dictating, snoozed }

  private weak var relay: RelayClient?
  private let wearables: any WearablesInterface
  private var session: DeviceSession?
  private var display: Display?
  private var current: AgentYank?
  private var pending: AgentYank?
  private var dictatingPartial = ""
  private var peekTimer: Task<Void, Never>?
  private var snoozeTimer: Task<Void, Never>?
  private(set) var state: State = .ambient

  private let peekTimeoutMs: UInt64 = 300_000
  private let snoozeMs:      UInt64 = 60_000

  init(wearables: any WearablesInterface, relay: RelayClient) {
    self.wearables = wearables
    self.relay = relay
  }

  func onIdle(_ yank: AgentYank) {
    if current?.thread == yank.thread,
       state == .peeking || state == .engaged || state == .followup || state == .dictating {
      if yank.awaiting == .permission {
        current = yank
        renderPeek()
      }
      return
    }
    if state == .ambient, !yank.lastUserInput.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
      return
    }
    yank(yank)
  }

  func yank(_ yank: AgentYank) {
    if (state == .peeking || state == .engaged || state == .followup),
       let cur = current?.thread, cur != yank.thread {
      pending = yank
      return
    }
    snoozeTimer?.cancel(); snoozeTimer = nil
    current = yank
    openSessionAndPeek()
  }

  func cancelIfFor(thread: String) {
    if pending?.thread == thread { pending = nil }
    NSLog("[face-chat] thread_busy %@ (HUD stays until dismissed)", thread)
  }

  private func openSessionAndPeek() {
    if display != nil {
      state = .peeking; renderPeek(); armPeekTimer(); return
    }
    do {
      let selector = AutoDeviceSelector(wearables: wearables, filter: { d in
        d.linkState == .connected && d.supportsDisplay()
      })
      let s = try wearables.createSession(deviceSelector: selector)
      session = s
      Task { [weak self] in
        for await st in s.stateStream() {
          guard let self else { return }
          if st == .started, self.display == nil { await self.attachDisplay(s) }
        }
      }
      try s.start()
    } catch {
      NSLog("[face-chat] createSession failed: \(error)")
    }
  }

  private func attachDisplay(_ s: DeviceSession) async {
    do {
      let cap = try s.addDisplay()
      display = cap
      _ = cap.statePublisher.listen { [weak self] ds in
        Task { @MainActor in
          guard let self else { return }
          if ds == .started {
            self.state = .peeking
            self.renderPeek()
            self.armPeekTimer()
          }
        }
      }
      await cap.start()
    } catch {
      NSLog("[face-chat] addDisplay failed: \(error)")
    }
  }

  private func armPeekTimer() {
    peekTimer?.cancel()
    peekTimer = Task { [weak self] in
      guard let ns = self?.peekTimeoutMs else { return }
      try? await Task.sleep(nanoseconds: ns * 1_000_000)
      if self?.state == .peeking { self?.closeSession() }
    }
  }

  private func renderPeek() {
    guard let d = display, let y = current else { return }
    let chips = ChipSet.forYank(y).prefix(3)
    let buttons: [any ViewComponent] = chips.enumerated().map { i, c in
      let style: ButtonStyle = i == 0 ? .primary : .secondary
      return Button(label: c.label, style: style, onClick: { Task { @MainActor in self.onChip(c) } })
    }
    let view = FlexBox(direction: .column, spacing: 8) {
      Text(y.metaLine, style: .meta, color: .secondary)
      FlexBox(direction: .column) {
        Text(String(y.bodyText.prefix(220)), style: .body)
      }
      .padding(12)
      .background(.card)
      FlexBox(direction: .row, spacing: 6, wrap: true) { for b in buttons { b } }
    }
    Task { try? await d.send(view) }
  }

  private func renderExpanded() {
    guard let d = display, let y = current else { return }
    let chips = ChipSet.forYank(y)
    let buttons: [any ViewComponent] = chips.prefix(3).enumerated().map { i, c in
      let style: ButtonStyle = i == 0 ? .primary : .secondary
      return Button(label: c.label, style: style, onClick: { Task { @MainActor in self.onChip(c) } })
    }
    let view = FlexBox(direction: .column, spacing: 8) {
      Text(y.metaLine, style: .meta, color: .secondary)
      FlexBox(direction: .column) { Text(y.bodyText, style: .body) }
        .padding(12).background(.card)
      FlexBox(direction: .row, spacing: 6, wrap: true) { for b in buttons { b } }
    }
    Task { try? await d.send(view) }
  }

  private func renderFollowUp() {
    guard let d = display, let y = current else { return }
    state = .followup
    peekTimer?.cancel()
    let chips = ChipSet.followUpChips(agent: y.agent)
    let buttons: [any ViewComponent] = chips.map { c in
      Button(label: c.label, style: .primary, onClick: { Task { @MainActor in self.onChip(c) } })
    }
    let view = FlexBox(direction: .column, spacing: 8) {
      Text("\(y.label) · follow-up", style: .meta, color: .secondary)
      FlexBox(direction: .column) { Text("pick a reply to send", style: .body) }
        .padding(12).background(.card)
      FlexBox(direction: .row, spacing: 6, wrap: true) {
        for b in buttons { b }
        Button(label: "back", style: .outline, onClick: { Task { @MainActor in self.renderExpanded() } })
      }
    }
    Task { try? await d.send(view) }
  }

  private func onEngage() { state = .engaged; peekTimer?.cancel(); renderExpanded() }
  private func onSnooze() {
    state = .snoozed
    let y = current
    closeSession(clearPending: false)
    snoozeTimer = Task { [weak self] in
      guard let ms = self?.snoozeMs else { return }
      try? await Task.sleep(nanoseconds: ms * 1_000_000)
      if let y { self?.yank(y) }
    }
  }
  private func onChip(_ c: Chip) {
    let y = current
    switch c.kind {
    case .send:
      if let y, let text = c.text { relay?.sendInput(thread: y.thread, text: text, enter: c.enter) }
      closeSession()
    case .dictate:
      if let y { Task { await startDictating(y) } }
    case .modify:
      closeSession()
    case .snooze: onSnooze()
    }
  }

  private func startDictating(_ y: AgentYank) async {
    if DictationManager.shared.isActive() { return }
    state = .dictating
    dictatingPartial = ""
    peekTimer?.cancel()
    relay?.sendDictateBegin(thread: y.thread)
    renderDictating()
    let dm = DictationManager.shared
    dm.onPartial = { [weak self] text in
      Task { @MainActor in
        self?.dictatingPartial = text
        self?.relay?.sendDictatePartial(thread: y.thread, text: text)
        self?.renderDictating()
      }
    }
    dm.onFinal = { [weak self] text in
      Task { @MainActor in
        self?.relay?.sendDictateCommit(thread: y.thread, text: text)
        self?.closeSession()
      }
    }
    dm.onCancelled = { [weak self] in
      Task { @MainActor in
        self?.relay?.sendDictateAbort(thread: y.thread)
        self?.state = .peeking
        self?.renderPeek()
        self?.armPeekTimer()
      }
    }
    dm.onError = { [weak self] msg in
      Task { @MainActor in
        self?.relay?.sendDictateAbort(thread: y.thread)
        NSLog("[face-chat] dictate error: \(msg)")
        self?.closeSession()
      }
    }
    await dm.start()
  }

  private func renderDictating() {
    guard let d = display, let y = current else { return }
    let line = dictatingPartial.trimmingCharacters(in: .whitespacesAndNewlines)
    let body = line.isEmpty ? "listening… speak now" : String(line.prefix(220))
    let view = FlexBox(direction: .column, spacing: 8) {
      Text("dictating", style: .meta, color: .secondary)
      FlexBox(direction: .column) { Text(body, style: .body) }
        .padding(12).background(.card)
      FlexBox(direction: .row, spacing: 6, wrap: true) {
        Button(label: "send", style: .primary, onClick: { Task { @MainActor in
          DictationManager.shared.stop(commit: true)
        }})
        Button(label: "cancel", style: .secondary, onClick: { Task { @MainActor in
          DictationManager.shared.stop(commit: false)
        }})
      }
    }
    Task { try? await d.send(view) }
  }

  private func closeSession(clearPending: Bool = true) {
    if DictationManager.shared.isActive() {
      DictationManager.shared.stop(commit: false)
    }
    dictatingPartial = ""
    peekTimer?.cancel(); peekTimer = nil
    session?.stop()
    display = nil
    session = nil
    current = nil
    state = .ambient
    let next = clearPending ? pending : nil
    if clearPending { pending = nil }
    if let next { yank(next) }
  }
}

File: relay-android/app/src/main/java/com/lowkey/facechat/hud/HudPresenter.kt
package com.lowkey.facechat.hud

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import com.lowkey.facechat.dictation.DictationCallback
import com.lowkey.facechat.dictation.DictationManager
import com.lowkey.facechat.relay.RelayClient
import com.lowkey.facechat.wearables.WearablesRepository
import com.lowkey.facechat.wearables.WearablesRuntime
import com.meta.wearable.dat.core.types.DeviceIdentifier
import com.meta.wearable.dat.core.types.LinkState
import com.meta.wearable.dat.display.Display
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeoutOrNull

class HudPresenter(
  private val appContext: Context,
  private val relay: RelayClient,
  private val wearables: WearablesRepository,
) {
  enum class State { AMBIENT, PEEKING, ENGAGED, FOLLOWUP, DICTATING, SNOOZED }

  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
  private val datSession = DatDisplaySession(scope)
  private var current: AgentYank? = null
  private var pending: AgentYank? = null
  private var dictatingPartial = ""
  private var peekTimer: Job? = null
  private var snoozeTimer: Job? = null
  private var state: State = State.AMBIENT

  private val PEEK_TIMEOUT_MS = 30_000L
  private val DICTATE_TIMEOUT_MS = 300_000L
  private val SNOOZE_MS = 60_000L
  private val LINK_WAIT_MS = 20_000L

  fun yank(yank: AgentYank) {
    if (!WearablesRuntime.initialized) {
      Log.w("HudPresenter", "Wearables SDK not initialized (open app and grant BT permissions)")
      return
    }
    if ((state == State.PEEKING || state == State.ENGAGED || state == State.FOLLOWUP) &&
      current?.thread != null && current?.thread != yank.thread) {
      pending = yank
      Log.i("HudPresenter", "queued yank thread=${yank.thread} (busy with ${current?.thread})")
      return
    }
    snoozeTimer?.cancel(); snoozeTimer = null
    current = yank
    openSessionAndPeek()
  }

  fun refresh(yank: AgentYank) {
    if (current?.thread == yank.thread &&
      (state == State.PEEKING || state == State.ENGAGED || state == State.FOLLOWUP)) {
      Log.i("HudPresenter", "refresh thread=${yank.thread} state=$state")
      current = yank
      renderCurrent()
      return
    }
    // Host echoes our reply as thread_idle — don't pop the card back up.
    if (state == State.AMBIENT && yank.lastUserInput.isNotBlank()) {
      Log.i("HudPresenter", "skip re-yank after user reply thread=${yank.thread}")
      return
    }
    yank(yank)
  }

  fun cancelIfFor(thread: String) {
    if (current?.thread == thread) closeSession()
    if (pending?.thread == thread) pending = null
  }

  private fun isHudDevice(d: com.meta.wearable.dat.core.types.Device): Boolean =
    d.isDisplayCapable()

  private fun connectedDisplayDevice(): DeviceIdentifier? {
    val meta = wearables.devicesMetadata.value
    return wearables.devices.value.firstOrNull { id ->
      val d = meta[id]
      d != null && isHudDevice(d) && d.linkState == LinkState.CONNECTED
    }
  }

  private fun displayCapableDevice(): DeviceIdentifier? {
    val meta = wearables.devicesMetadata.value
    return wearables.devices.value.firstOrNull { id ->
      val d = meta[id]
      d != null && isHudDevice(d)
    }
  }

  private fun openSessionAndPeek() {
    if (datSession.activeDisplay != null) {
      state = State.PEEKING
      renderPeek()
      armPeekTimer()
      return
    }

    val connected = connectedDisplayDevice()
    if (connected != null) {
      prepareAndPeek(connected)
      return
    }

    val candidate = displayCapableDevice()
    if (candidate == null) {
      Log.w("HudPresenter", "no display-capable device known to SDK — waiting for pairing")
      waitForDisplayDeviceThenPeek()
      return
    }

    scope.launch {
      val ok = withTimeoutOrNull(LINK_WAIT_MS) {
        wearables.devicesMetadata.first { meta ->
          meta[candidate]?.linkState == LinkState.CONNECTED
        }
      } != null
      if (ok) prepareAndPeek(candidate)
      else {
        Log.w("HudPresenter", "timed out waiting for CONNECTED — retrying when link appears")
        waitForDisplayDeviceThenPeek()
      }
    }
  }

  private fun waitForDisplayDeviceThenPeek() {
    scope.launch {
      withTimeoutOrNull(LINK_WAIT_MS * 3) {
        wearables.devicesMetadata.first { meta ->
          meta.values.any { isHudDevice(it) && it.linkState == LinkState.CONNECTED }
        }
      } ?: run {
        Log.w("HudPresenter", "still no connected display device after wait")
        return@launch
      }
      val id = connectedDisplayDevice() ?: return@launch
      prepareAndPeek(id)
    }
  }

  private fun prepareAndPeek(deviceId: DeviceIdentifier) {
    datSession.prepareDisplay(deviceId) { d ->
      state = State.PEEKING
      renderPeek(d)
      armPeekTimer()
    }
  }

  private fun armPeekTimer() {
    peekTimer?.cancel()
    peekTimer = scope.launch {
      delay(PEEK_TIMEOUT_MS)
      if (state == State.PEEKING) closeSession()
    }
  }

  private fun renderCurrent() {
    when (state) {
      State.PEEKING -> renderPeek()
      State.ENGAGED -> renderExpanded()
      State.FOLLOWUP -> renderFollowUp()
      State.DICTATING -> renderDictating()
      else -> {}
    }
  }

  private fun renderPeek(d: Display? = datSession.activeDisplay) {
    val display = d ?: return
    val y = current ?: return
    val chips = ChipSet.forYank(y)
    val primary = chips.firstOrNull() ?: return
    val secondary = chips.getOrNull(1) ?: return

    HudWidgets.sendPeek(
      scope, display, y,
      onPrimary = { onChip(primary) },
      onSecondary = { onChip(secondary) },
      primaryLabel = primary.label,
      secondaryLabel = secondary.label,
    )
  }

  private fun renderExpanded() {
    val d = datSession.activeDisplay ?: return
    val y = current ?: return
    HudWidgets.sendExpanded(scope, d, y, ChipSet.forYank(y), ::onChip)
  }

  private fun renderFollowUp() {
    val d = datSession.activeDisplay ?: return
    val y = current ?: return
    state = State.FOLLOWUP
    peekTimer?.cancel()
    HudWidgets.sendFollowUp(scope, d, y, ChipSet.followUpChips(y.agent), ::onChip)
  }

  private fun renderDictating() {
    val d = datSession.activeDisplay ?: return
    val y = current ?: return
    HudWidgets.sendDictating(
      scope, d, y, dictatingPartial,
      onSend = { finishDictating(commit = true) },
      onCancel = { finishDictating(commit = false) },
    )
  }

  private fun finishDictating(commit: Boolean) {
    val y = current ?: return
    if (commit) {
      DictationManager.stop(commitPartial = true)
    } else {
      DictationManager.stop(commitPartial = false)
      relay.sendDictateAbort(y.thread)
      dictatingPartial = ""
      state = State.PEEKING
      renderPeek()
      armPeekTimer()
    }
  }

  private fun startDictating(y: AgentYank) {
    if (DictationManager.isActive()) return
    state = State.DICTATING
    dictatingPartial = ""
    peekTimer?.cancel()
    peekTimer = scope.launch {
      delay(DICTATE_TIMEOUT_MS)
      if (state == State.DICTATING) finishDictating(commit = true)
    }
    relay.sendDictateBegin(y.thread)
    renderDictating()
    DictationManager.start(
      appContext,
      object : DictationCallback {
        override fun onPartial(text: String) {
          dictatingPartial = text
          relay.sendDictatePartial(y.thread, text)
          renderDictating()
        }
        override fun onFinal(text: String) {
          dictatingPartial = ""
          relay.sendDictateCommit(y.thread, text)
          DictationManager.stop(commitPartial = false, notify = false)
          closeSession()
        }
        override fun onCancelled() {
          relay.sendDictateAbort(y.thread)
          dictatingPartial = ""
          state = State.PEEKING
          renderPeek()
          armPeekTimer()
        }
        override fun onError(message: String) {
          Log.w("HudPresenter", "dictate error: $message")
          relay.sendDictateAbort(y.thread)
          dictatingPartial = ""
          state = State.PEEKING
          renderPeek()
          armPeekTimer()
        }
      },
    )
  }

  private fun onChip(c: Chip) {
    val y = current ?: return
    if (state == State.DICTATING) {
      when (c.label) {
        "send" -> finishDictating(commit = true)
        "cancel" -> finishDictating(commit = false)
      }
      return
    }
    Log.i("HudPresenter", "chip tapped: ${c.label} thread=${y.thread}")
    when (c.kind) {
      ChipKind.SEND -> {
        if (c.text != null) {
          relay.sendInput(y.thread, c.text, c.enter)
          closeSession()
        }
      }
      ChipKind.DICTATE -> startDictating(y)
      ChipKind.MODIFY -> {
        try {
          val url = relay.companionComposeUrl(y.thread)
          Log.i("HudPresenter", "modify → companion $url")
          appContext.startActivity(
            Intent(Intent.ACTION_VIEW, Uri.parse(url)).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK),
          )
        } catch (e: Exception) {
          Log.e("HudPresenter", "modify open failed: ${e.message}")
        }
        closeSession()
      }
      ChipKind.DISMISS -> closeSession()
      ChipKind.SNOOZE -> onSnooze()
    }
  }

  private fun onSnooze() {
    state = State.SNOOZED
    val y = current
    closeSession(clearPending = false)
    snoozeTimer = scope.launch {
      delay(SNOOZE_MS)
      if (y != null) yank(y)
    }
  }

  private fun closeSession(clearPending: Boolean = true) {
    peekTimer?.cancel(); peekTimer = null
    if (DictationManager.isActive()) {
      DictationManager.stop(commitPartial = false, notify = false)
    }
    dictatingPartial = ""
    datSession.stop()
    current = null
    state = State.AMBIENT
    val next = if (clearPending) pending.also { pending = null } else null
    if (next != null) yank(next)
  }
}
